src/annotator/guest.js

```diff
@@ -421,98 +402,62 @@ export default class Guest {
-
-      // Let other frames (eg. the sidebar) know about the new annotation.
-      this.crossframe.sync([annotation]);
-
-      return anchors;
-    };
+    // Remove existing anchors for this annotation.
+    this.detach(annotation, false /* notify */);
 
-    // Remove all the anchors for this annotation from the instance storage.
-    for (anchor of this.anchors.splice(0, this.anchors.length)) {
-      if (anchor.annotation === annotation) {
-        // Anchors are valid as long as they still have a range and their target
-        // is still in the list of targets for this annotation.
-        if (anchor.range && annotation.target.includes(anchor.target)) {
-          anchors.push(anchor);
-          anchoredTargets.push(anchor.target);
-        } else if (anchor.highlights) {
-          // These highlights are no longer valid and should be removed.
-          deadHighlights = deadHighlights.concat(anchor.highlights);
-          delete anchor.highlights;
-          delete anchor.range;
-        }
-      } else {
-        // These can be ignored, so push them back onto the new list.
-        this.anchors.push(anchor);
-      }
+    // Resolve selectors to ranges and insert highlights.
+    if (!annotation.target) {
+      annotation.target = [];
+    }
+    const anchors = await Promise.all(annotation.target.map(locate));
+    for (let anchor of anchors) {
+      highlight(anchor);
     }
 
-    // Remove all the highlights that have no corresponding target anymore.
-    removeHighlights(deadHighlights);
+    // Set flag indicating whether anchoring succeeded. For each target,
+    // anchoring is successful either if there are no selectors (ie. this is a
+    // Page Note) or we successfully resolved the selectors to a range.
+    annotation.$orphan =
+      anchors.length > 0 &&
+      anchors.every(anchor => anchor.target.selector && !anchor.range);
 
-    // Anchor any targets of this annotation that are not anchored already.
-    for (let target of annotation.target) {
-      if (!anchoredTargets.includes(target)) {
-        anchor = locate(target).then(highlight);
-        anchors.push(anchor);
-      }
-    }
+    this._updateAnchors(this.anchors.concat(anchors), true /* notify */);
+
+    // Let other frames (eg. the sidebar) know about the new annotation.
+    this.crossframe.sync([annotation]);
 
-    return Promise.all(anchors).then(sync);
+    return anchors;
   }
 
   /**
    * Remove the anchors and associated highlights for an annotation from the document.
    *
    * @param {AnnotationData} annotation
+   * @param {boolean} [notify] - For internal use. Whether to emit an `anchorsChanged` notification
    */
-  detach(annotation) {
+  detach(annotation, notify = true) {
     const anchors = [];
-    let unhighlight = [];
     for (let anchor of this.anchors) {
-      if (anchor.annotation === annotation) {
-        unhighlight.push(...(anchor.highlights ?? []));
-      } else {
+      if (anchor.annotation !== annotation) {
```

Comment:
I don't see a test case where the annotation references are not equal.
e.g. we create 2 annotations, but don't push one to anchors, and then try and delete it.

Reply:
Indeed! - I've added a test to cover this and also refactored the existing `destroy` tests a little.